source/simulator/assembly.cc

```diff
@@ -1146,14 +1151,18 @@ namespace aspect
               // We only need to look up values of shape functions if they
               // belong to 'our' component. They are zero otherwise anyway.
               // Note that we later only look at the values that we do set here.
-              for (unsigned int k=0; k<advection_dofs_per_cell; ++k)
+              for (unsigned int k=0, i=0; i<advection_dofs_per_cell; ++k)
```

Comment:
You're right with the loop -- not necessary.

Moving the increment of `k` to the end of the loop works for me. Maybe put something like `/* loop counter increments at the end of the loop body */` into the third slot of the `for` header?